File: route_planner/excel_reader.py
import math
import re
from dataclasses import dataclass

import numpy as np
import pandas as pd

from .models import Stop


def _normalize_column(name: str) -> str:
    return re.sub(r"[\s_]+", " ", str(name).strip().lower())


def _find_column(columns: list[str], *candidates: str) -> str | None:
    normalized = {_normalize_column(c): c for c in columns}
    for candidate in candidates:
        key = _normalize_column(candidate)
        if key in normalized:
            return normalized[key]
    return None


def _coerce_float(value) -> float | None:
    if pd.isna(value):
        return None
    try:
        return float(value)
    except (TypeError, ValueError):
        return None


@dataclass
class ColumnMapping:
    lat: str
    lng: str
    name: str | None
    code: str | None
    start_lat: str | None
    start_lng: str | None
    district: str | None
    tehsil: str | None


def resolve_columns(columns: list[str]) -> ColumnMapping:
    lat_col = _find_column(
        columns,
        "Latitude",
        "Lat",
        "Dest Latitude",
        "Destination Latitude",
        "Y",
    )
    lng_col = _find_column(
        columns,
        "Longitude",
        "Lng",
        "Long",
        "Dest Longitude",
        "Destination Longitude",
        "X",
    )
    if lat_col is None or lng_col is None:
        raise ValueError(
            "Excel file must contain latitude and longitude columns "
            "(e.g. Latitude / Longitude)."
        )
    return ColumnMapping(
        lat=lat_col,
        lng=lng_col,
        name=_find_column(
            columns,
            "Name",
            "Location Name",
            "Place Name",
            "Site Name",
            "Destination",
            "Destination Name",
            "School Name",
            "Stop Name",
        ),
        code=_find_column(
            columns,
            "SchoolCode",
            "School Code",
            "Code",
            "Location Code",
            "ID",
            "Ref",
        ),
        start_lat=_find_column(
            columns,
            "Start LAT",
            "Start Lat",
            "Start Latitude",
            "Origin Lat",
            "Origin Latitude",
            "Depot Lat",
        ),
        start_lng=_find_column(
            columns,
            "Start LNG",
            "Start Lng",
            "Start Longitude",
            "Origin Lng",
            "Origin Longitude",
            "Depot Lng",
        ),
        district=_find_column(columns, "District", "Region", "Area"),
        tehsil=_find_column(columns, "Tehsil", "Tehsil Name", "Sub District", "Sub-District"),
    )


def read_route_dataframe(path: str) -> tuple[pd.DataFrame, ColumnMapping]:
    """Load Excel and resolve column mapping (generic — any compatible format)."""
    df = pd.read_excel(path)
    mapping = resolve_columns(list(df.columns))
    return df, mapping


def read_school_dataframe(path: str) -> tuple[pd.DataFrame, ColumnMapping]:
    """Backward-compatible alias for read_route_dataframe."""
    return read_route_dataframe(path)


def read_start_coordinates(
    df: pd.DataFrame,
    mapping: ColumnMapping,
    start_lat: float | None = None,
    start_lng: float | None = None,
) -> tuple[float, float]:
    if start_lat is not None and start_lng is not None:
        return start_lat, start_lng
    if mapping.start_lat and mapping.start_lng:
        lat = _coerce_float(df[mapping.start_lat].dropna().iloc[0])
        lng = _coerce_float(df[mapping.start_lng].dropna().iloc[0])
        if lat is not None and lng is not None:
            return lat, lng
    raise ValueError(
        "Start coordinates not found. Provide --start-lat/--start-lng or "
        "include Start LAT / Start LNG columns in the Excel file."
    )


def dataframe_to_stops(
    df: pd.DataFrame,
    mapping: ColumnMapping,
    group_col: str | None = None,
    group_value: str | None = None,
) -> list[Stop]:
    work = df
    if group_col and group_value is not None:
        work = df[df[group_col].astype(str).str.strip() == str(group_value).strip()]

    destinations: list[Stop] = []
    seen: set[tuple[float, float]] = set()

    for idx, row in work.iterrows():
        lat = _coerce_float(row[mapping.lat])
        lng = _coerce_float(row[mapping.lng])
        if lat is None or lng is None:
            continue

        key = (round(lat, 6), round(lng, 6))
        if key in seen:
            continue
        seen.add(key)

        name = (
            str(row[mapping.name]).strip()
            if mapping.name and pd.notna(row[mapping.name])
            else f"Stop {len(destinations) + 1}"
        )
        code = str(row[mapping.code]).strip() if mapping.code and pd.notna(row[mapping.code]) else ""
        district = (
            str(row[mapping.district]).strip()
            if mapping.district and pd.notna(row[mapping.district])
            else ""
        )
        tehsil = str(row[mapping.tehsil]).strip() if mapping.tehsil and pd.notna(row[mapping.tehsil]) else ""

        destinations.append(
            Stop(
                name=name,
                lat=lat,
                lng=lng,
                school_code=code,
                original_index=int(idx),
                district=district,
                tehsil=tehsil,
            )
        )

    return destinations


def list_groups(df: pd.DataFrame, mapping: ColumnMapping, group_by: str) -> list[str]:
    col = _group_column(mapping, group_by)
    if col is None:
        return ["All Locations"]
    values = df[col].dropna().astype(str).str.strip()
    values = values[values != ""]
    return sorted(values.unique().tolist())


def _group_column(mapping: ColumnMapping, group_by: str) -> str | None:
    if group_by == "district":
        return mapping.district
    if group_by == "tehsil":
        return mapping.tehsil
    return None


def split_stops_geographically(stops: list[Stop], max_per_batch: int) -> list[list[Stop]]:
    """Split a large stop list into geographic sub-batches using k-means."""
    if len(stops) <= max_per_batch:
        return [stops]

    n_batches = math.ceil(len(stops) / max_per_batch)
    coords = np.array([[s.lat, s.lng] for s in stops], dtype=np.float64)

    # Initialize centroids by spreading across latitude
    order = np.argsort(coords[:, 0])
    step = max(1, len(order) // n_batches)
    centroid_idx = [order[min(i * step, len(order) - 1)] for i in range(n_batches)]
    centroids = coords[centroid_idx].copy()

    assignments = np.zeros(len(stops), dtype=int)
    for _ in range(20):
        dists = np.linalg.norm(coords[:, None, :] - centroids[None, :, :], axis=2)
        assignments = np.argmin(dists, axis=1)
        new_centroids = np.zeros_like(centroids)
        for k in range(n_batches):
            mask = assignments == k
            if mask.any():
                new_centroids[k] = coords[mask].mean(axis=0)
            else:
                new_centroids[k] = centroids[k]
        if np.allclose(new_centroids, centroids):
            break
        centroids = new_centroids

    batches: list[list[Stop]] = [[] for _ in range(n_batches)]
    for stop, cluster in zip(stops, assignments):
        batches[cluster].append(stop)

    # Rebalance oversized batches
    result: list[list[Stop]] = []
    overflow: list[Stop] = []
    for batch in batches:
        if len(batch) > max_per_batch:
            overflow.extend(batch)
        elif batch:
            result.append(batch)

    while overflow:
        chunk = overflow[:max_per_batch]
        overflow = overflow[max_per_batch:]
        if chunk:
            result.append(chunk)

    return result if result else [stops]


def load_stops_from_excel(
    path: str,
    start_lat: float | None = None,
    start_lng: float | None = None,
    start_name: str = "Start",
    group_by: str | None = None,
    group_value: str | None = None,
) -> tuple[Stop, list[Stop]]:
    """Load start point and destinations from an Excel file."""
    df, mapping = read_route_dataframe(path)
    lat, lng = read_start_coordinates(df, mapping, start_lat, start_lng)
    start = Stop(name=start_name, lat=lat, lng=lng)

    group_col = _group_column(mapping, group_by) if group_by else None
    destinations = dataframe_to_stops(df, mapping, group_col, group_value)

    if not destinations:
        label = group_value or "file"
        raise ValueError(f"No valid destination coordinates found for {label}.")

    return start, destinations

File: src/RoutePlanner.Core/Excel/ExcelReader.cs
using System.Text.RegularExpressions;
using ClosedXML.Excel;
using RoutePlanner.Core.Models;

namespace RoutePlanner.Core.Excel;

public sealed class ColumnMapping
{
    public required int LatCol { get; init; }
    public required int LngCol { get; init; }
    public int? NameCol { get; init; }
    public int? CodeCol { get; init; }
    public int? StartLatCol { get; init; }
    public int? StartLngCol { get; init; }
    public int? DistrictCol { get; init; }
    public int? TehsilCol { get; init; }
}

public sealed class RouteExcelData : IDisposable
{
    private readonly XLWorkbook _workbook;

    public RouteExcelData(string path)
    {
        _workbook = new XLWorkbook(path);
        Sheet = _workbook.Worksheet(1);
        var headerCells = Sheet.Row(1).CellsUsed().ToList();
        var headers = headerCells.Select(c => c.GetString()).ToList();
        Mapping = ColumnResolver.Resolve(headers, headerCells);
    }

    public IXLWorksheet Sheet { get; }
    public ColumnMapping Mapping { get; }
    public void Dispose() => _workbook.Dispose();
}

public static class ExcelReader
{
    public static RouteExcelData Open(string path) => new(path);

    public static (double Lat, double Lng) ReadStartCoordinates(
        RouteExcelData data,
        double? overrideLat = null,
        double? overrideLng = null)
    {
        if (overrideLat.HasValue && overrideLng.HasValue)
            return (overrideLat.Value, overrideLng.Value);

        var m = data.Mapping;
        if (m.StartLatCol is null || m.StartLngCol is null)
            throw new InvalidOperationException(
                "Start coordinates not found. Include Start LAT / Start LNG columns or provide them manually.");

        foreach (var row in data.Sheet.RowsUsed().Skip(1))
        {
            var lat = CoerceDouble(row.Cell(m.StartLatCol.Value).GetString());
            var lng = CoerceDouble(row.Cell(m.StartLngCol.Value).GetString());
            if (lat.HasValue && lng.HasValue)
                return (lat.Value, lng.Value);
        }

        throw new InvalidOperationException("Start LAT / Start LNG columns exist but contain no valid values.");
    }

    public static List<Stop> ToStops(RouteExcelData data, int? groupColumn = null, string? groupValue = null)
    {
        var m = data.Mapping;
        var stops = new List<Stop>();
        var seen = new HashSet<(double, double)>();

        foreach (var row in data.Sheet.RowsUsed().Skip(1))
        {
            if (groupColumn.HasValue && groupValue is not null)
            {
                var gv = row.Cell(groupColumn.Value).GetString().Trim();
                if (!string.Equals(gv, groupValue.Trim(), StringComparison.OrdinalIgnoreCase))
                    continue;
            }

            var lat = CoerceDouble(row.Cell(m.LatCol).GetString());
            var lng = CoerceDouble(row.Cell(m.LngCol).GetString());
            if (!lat.HasValue || !lng.HasValue) continue;

            var key = (Math.Round(lat.Value, 6), Math.Round(lng.Value, 6));
            if (!seen.Add(key)) continue;

            var name = m.NameCol.HasValue ? row.Cell(m.NameCol.Value).GetString().Trim() : "";
            if (string.IsNullOrWhiteSpace(name)) name = $"Stop {stops.Count + 1}";

            stops.Add(new Stop
            {
                Name = name,
                Lat = lat.Value,
                Lng = lng.Value,
                LocationCode = m.CodeCol.HasValue ? row.Cell(m.CodeCol.Value).GetString().Trim() : "",
                District = m.DistrictCol.HasValue ? row.Cell(m.DistrictCol.Value).GetString().Trim() : "",
                Tehsil = m.TehsilCol.HasValue ? row.Cell(m.TehsilCol.Value).GetString().Trim() : "",
                OriginalIndex = row.RowNumber()
            });
        }

        return stops;
    }

    public static List<string> ListGroups(RouteExcelData data, string groupBy)
    {
        var col = GroupColumn(data.Mapping, groupBy);
        if (col is null) return ["All Locations"];

        return data.Sheet.RowsUsed().Skip(1)
            .Select(r => r.Cell(col.Value).GetString().Trim())
            .Where(v => !string.IsNullOrWhiteSpace(v))
            .Distinct(StringComparer.OrdinalIgnoreCase)
            .OrderBy(v => v, StringComparer.OrdinalIgnoreCase)
            .ToList();
    }

    public static int? GroupColumn(ColumnMapping mapping, string groupBy) => groupBy switch
    {
        "district" => mapping.DistrictCol,
        "tehsil" => mapping.TehsilCol,
        _ => null
    };

    public static void WriteTemplate(string path)
    {
        using var wb = new XLWorkbook();
        var ws = wb.AddWorksheet("Locations");
        ws.Cell(1, 1).Value = "SchoolCode";
        ws.Cell(1, 2).Value = "Name";
        ws.Cell(1, 3).Value = "District";
        ws.Cell(1, 4).Value = "Tehsil";
        ws.Cell(1, 5).Value = "Latitude";
        ws.Cell(1, 6).Value = "Longitude";
        ws.Cell(1, 7).Value = "Start LAT";
        ws.Cell(1, 8).Value = "Start LNG";
        ws.Cell(2, 1).Value = "LOC-001";
        ws.Cell(2, 2).Value = "Example Location A";
        ws.Cell(2, 3).Value = "DISTRICT_A";
        ws.Cell(2, 4).Value = "TEHSIL_1";
        ws.Cell(2, 5).Value = 31.5204;
        ws.Cell(2, 6).Value = 74.3587;
        ws.Cell(2, 7).Value = 31.5354;
        ws.Cell(2, 8).Value = 74.3442;
        ws.Cell(3, 1).Value = "LOC-002";
        ws.Cell(3, 2).Value = "Example Location B";
        ws.Cell(3, 3).Value = "DISTRICT_A";
        ws.Cell(3, 4).Value = "TEHSIL_1";
        ws.Cell(3, 5).Value = 31.5497;
        ws.Cell(3, 6).Value = 74.3436;
        ws.Cell(3, 7).Value = 31.5354;
        ws.Cell(3, 8).Value = 74.3442;
        wb.SaveAs(path);
    }

    private static double? CoerceDouble(string value)
    {
        if (string.IsNullOrWhiteSpace(value)) return null;
        return double.TryParse(value, out var d) ? d : null;
    }
}

internal static class ColumnResolver
{
    public static ColumnMapping Resolve(IReadOnlyList<string> headers, IReadOnlyList<IXLCell> headerCells)
    {
        int ColIndex(string name) => headerCells.First(c =>
            string.Equals(c.GetString().Trim(), name, StringComparison.OrdinalIgnoreCase)).Address.ColumnNumber;

        var latName = FindColumn(headers, "Latitude", "Lat", "Dest Latitude", "Destination Latitude", "Y")
                      ?? throw new InvalidOperationException("Excel must contain a Latitude column.");
        var lngName = FindColumn(headers, "Longitude", "Lng", "Long", "Dest Longitude", "Destination Longitude", "X")
                      ?? throw new InvalidOperationException("Excel must contain a Longitude column.");

        int? OptionalCol(params string[] candidates)
        {
            var name = FindColumn(headers, candidates);
            return name is null ? null : ColIndex(name);
        }

        return new ColumnMapping
        {
            LatCol = ColIndex(latName),
            LngCol = ColIndex(lngName),
            NameCol = OptionalCol("Name", "Location Name", "Place Name", "Site Name", "Destination", "Destination Name", "School Name", "Stop Name"),
            CodeCol = OptionalCol("SchoolCode", "School Code", "Code", "Location Code", "ID", "Ref"),
            StartLatCol = OptionalCol("Start LAT", "Start Lat", "Start Latitude", "Origin Lat", "Origin Latitude", "Depot Lat"),
            StartLngCol = OptionalCol("Start LNG", "Start Lng", "Start Longitude", "Origin Lng", "Origin Longitude", "Depot Lng"),
            DistrictCol = OptionalCol("District", "Region", "Area"),
            TehsilCol = OptionalCol("Tehsil", "Tehsil Name", "Sub District", "Sub-District")
        };
    }

    private static string? FindColumn(IReadOnlyList<string> columns, params string[] candidates)
    {
        var normalized = columns.ToDictionary(c => Normalize(c), c => c, StringComparer.OrdinalIgnoreCase);
        foreach (var candidate in candidates)
        {
            if (normalized.TryGetValue(Normalize(candidate), out var col))
                return col;
        }
        return null;
    }

    private static string Normalize(string name) =>
        Regex.Replace(name.Trim(), @"\s+", " ").ToLowerInvariant();
}
